df.history.xml

```diff
@@ -2200,6 +2222,203 @@
+      <int32_t name='hf' ref-target='historical_figure'/>
+      <int32_t name='site' ref-target='world_site'/>
+      <int16_t name='content' comment='11 = inveighing against, 12 = urging love be shown'/>
+      <int32_t name='preached_to_entity' ref-target='historical_entity'/>
+      <int32_t name='against_entity' ref-target='historical_entity'/>
+    </class-type>
+
+    <class-type type-name='history_event_entity_persecutedst' inherits-from='history_event' since='v0.47.01'>
+      <int32_t name='persecuting_hf' ref-target='historical_figure'/>
+      <int32_t name='persecuting_entity' ref-target='historical_entity'/>
+      <int32_t name='target_entity' ref-target='historical_entity'/>
+      <int32_t name='site' ref-target='world_site'/>
+      <stl-vector type-name='int32_t' ref-target='historical_figure' comment='subset of expelled_hfs. expelled priests?'/>
+      <stl-vector type-name='int32_t'/>
+      <int32_t/>
+      <stl-vector name='expelled_hfs' type-name='int32_t' ref-target='historical_figure'/>
+      <stl-vector type-name='int32_t'/>
+      <stl-vector name='expelled_races' type-name='int32_t'/>
+      <stl-vector name='expelled_counts' type-name='int32_t'/>
+    </class-type>
+
+    <class-type type-name='history_event_entity_breach_feature_layerst' inherits-from='history_event' since='v0.47.01'>
+      <int32_t name='site' ref-target='world_site'/>
+      <int32_t name='governing_entity' ref-target='historical_entity'/>
+      <int32_t name='civ_entity' ref-target='historical_entity'/>
+      <int32_t name='layer' ref-target='world_underground_region'/>
+    </class-type>
+
+    <class-type type-name='history_event_entity_alliance_formedst' inherits-from='history_event' since='v0.47.01'>
+      <int32_t name='entity' ref-target='historical_entity'/>
+      <int32_t/>
+      <stl-vector name='allied_entities' type-name='int32_t' ref-target='historical_entity'/>
+    </class-type>
+
+    <class-type type-name='history_event_hf_ransomedst' inherits-from='history_event' since='v0.47.01'>
+      <int32_t name='hf' ref-target='historical_figure'/>
+      <int32_t name='ransomer_hf' ref-target='historical_figure'/>
+      <int32_t name='paying_hf' ref-target='historical_figure'/>
+      <int32_t name='paying_entity' ref-target='historical_entity'/>
+      <int32_t name='site' ref-target='world_site'/>
+    </class-type>
+
+    <class-type type-name='history_event_hf_enslavedst' inherits-from='history_event' since='v0.47.01'>
+      <int32_t name='hf' ref-target='historical_figure'/>
+      <int32_t name='selling_hf' ref-target='historical_figure'/>
+      <int32_t name='buying_entity' ref-target='historical_entity'/>
+      <int32_t name='site' ref-target='world_site'/>
+    </class-type>
+
+    <class-type type-name='history_event_sabotagest' inherits-from='history_event' since='v0.47.01'>
+       Have not yet found this event in-game
```

Comment:
How long of a history have you tried? A general rule of thumb is longer histories -> more varied events. (Of course, a 2500 year history would be a bit of a pain, so don't overdo it; if this is hard to find, we can leave it for now)

Reply:
I did a couple of medium worlds at ~600 years.  I'll try one a little longer, but I'm starting to think that this event type is not in use--maybe it got delayed?  At first I thought it was going to be an event that only showed up in fortress mode, but when validating structure sizes, it looks like the `history_event_sabotagest` struct doesn't have any fields of its own, i.e. it's the same size as its parent class.  I'm not entirely sure if my methodology for figuring out struct sizes is correct.

Reply:
I've just checked this class against version 0.47.02 and its layout is as follows:
```
[base class]
int32 saboteur_hfid;
int32 target_hfid;
int32 target_enid;
int32 site_id;
```
The structure size was gathered by finding code that instantiated it (searching for vtable xrefs), the field layout was determined by examining the read_file/write_file vmethods, and the field names were determined by examining the generate_xml vmethod.

Reply:
Thanks for describing how you did that--I'm definitely still learning how all of this fits together.  I did not know that you could take that approach.